core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -398,27 +386,31 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
    *
    * If no format conversion or value overwriting is required for messages, this method will perform in-place
    * operations and avoid re-compression.
+   *
+   * Returns the message set and a boolean indicating whether recompression took place.
    */
-  private[kafka] def validateMessagesAndAssignOffsets(offsetCounter: AtomicLong,
-                                                      now: Long = System.currentTimeMillis(),
+  private[kafka] def validateMessagesAndAssignOffsets(offsetCounter: LongRef,
+                                                      now: Long,
                                                       sourceCodec: CompressionCodec,
                                                       targetCodec: CompressionCodec,
                                                       compactedTopic: Boolean = false,
                                                       messageFormatVersion: Byte = Message.CurrentMagicValue,
                                                       messageTimestampType: TimestampType,
-                                                      messageTimestampDiffMaxMs: Long): ByteBufferMessageSet = {
+                                                      messageTimestampDiffMaxMs: Long): (ByteBufferMessageSet, Boolean) = {
     if (sourceCodec == NoCompressionCodec && targetCodec == NoCompressionCodec) {
-      // check the magic value
-      if (!magicValueInAllWrapperMessages(messageFormatVersion)) {
-        // Message format conversion
-        convertNonCompressedMessages(offsetCounter, compactedTopic, now, messageTimestampType, messageTimestampDiffMaxMs,
-          messageFormatVersion)
-      } else {
-        // Do in-place validation, offset assignment and maybe set timestamp
-        validateNonCompressedMessagesAndAssignOffsetInPlace(offsetCounter, now, compactedTopic, messageTimestampType,
-          messageTimestampDiffMaxMs)
+      val messageSet = {
+        // check the magic value
+        if (!isMagicValueInAllWrapperMessages(messageFormatVersion)) {
+          // Message format conversion
+          convertNonCompressedMessages(offsetCounter, compactedTopic, now, messageTimestampType, messageTimestampDiffMaxMs,
```

Comment:
We probably should return true in this case since the message entrySize may now be bigger if we are converting from v0 message format to v1. We will need to change the comment on the return value accordingly.